src/python.cpp

```diff
@@ -603,63 +505,64 @@ std::string py_fetch_error() {
   // returned back to the top level.
   if (reticulate::signals::getPythonInterruptsPending()) {
     PyErr_Clear();
+    Py_DecRef(excValue);
     reticulate::signals::setInterruptsPending(false);
     reticulate::signals::setPythonInterruptsPending(false);
     throw Rcpp::internal::InterruptedException();
   }
 
-  // read and normalize error, exception
-  PyObject *excType, *excValue, *excTraceback;
-  PyErr_Fetch(&excType, &excValue, &excTraceback);
-  PyErr_NormalizeException(&excType, &excValue, &excTraceback);
+  Environment pkg_globals(
+      Environment::namespace_env("reticulate").get(".globals"));
+  pkg_globals.assign("py_last_exception", py_ref(excValue, false));
 
-  // create object pointers (so they're freed on scope exit)
-  PyObjectPtr pExcType(excType);
-  PyObjectPtr pExcValue(excValue);
-  PyObjectPtr pExcTraceback(excTraceback);
+  // invoke 'traceback.format_exception_only(<traceback>)'
+  PyObjectPtr tb_module(py_import("traceback"));
+  if (tb_module.is_null())
+    return "<unknown python exception, traceback module not found>";
+
+  PyObjectPtr format_exception_only(PyObject_GetAttrString(tb_module, "format_exception_only"));
+  if (format_exception_only.is_null())
+    return "<unknown python exception, traceback format fn not found>";
 
-  // if we don't have any error information, return early
-  if (pExcType.is_null() && pExcValue.is_null())
-    return "<unknown error>";
+  PyObjectPtr formatted(PyObject_CallFunctionObjArgs(format_exception_only, excType, excValue, NULL));
+  if (formatted.is_null())
+    return "<unknown python exception, traceback format fn returned NULL>";
 
   // build error text
   std::ostringstream oss;
 
-  // get exception type
-  std::string type = py_fetch_error_type(pExcType);
-
-  // set type if we had it
-  if (!type.empty()) {
-    s_lastError.setType(type);
-    oss << type << ": ";
-  }
-
-  // get exception value
-  std::string value = py_fetch_error_value(pExcValue);
-  if (!value.empty()) {
-    s_lastError.setValue(value);
-    oss << value;
-  }
-
-  // retrieve Python traceback
-  std::vector<std::string> traceback;
-  py_fetch_error_traceback(excTraceback, &traceback);
-  s_lastError.setTraceback(traceback);
-
-  if (traceback_enabled()) {
-    std::size_t n = traceback.size();
-    if (n > 0) {
-      oss << "\n\nDetailed traceback:\n";
-      for (std::size_t i = 0; i < n; i++)
-        oss << traceback[i];
-    }
-  }
+  for (Py_ssize_t i = 0, n = PyList_Size(formatted); i < n; i++)
+    oss << as_std_string(PyList_GetItem(formatted, i));
 
-  // get final error string
   std::string error = oss.str();
-  s_lastError.setMessage(error);
 
-  // return error
+  int max_msg_len(Rf_asInteger(Rf_GetOption1(Rf_install("warning.length"))));
```

Comment:
Out of an abundance of caution, I would recommend PROTECTing the result of `Rf_GetOption1` before calling `Rf_asInteger()`, then UNPROTECTing after.